Content.Shared/SS220/Trap/TrapSystem.cs

```diff
@@ -0,0 +1,161 @@
+            used: args.User)
+        {
+            BreakOnMove = true,
+            AttemptFrequency = AttemptFrequency.StartAndEnd,
+        };
+
+        var localUser = args.User;
+        if (ent.Comp.IsArmed)
+        {
+            verb.Text = Loc.GetString("trap-component-defuse-trap");
+            verb.Act = () => _doAfter.TryStartDoAfter(doAfterArgs);
+        }
+        else
+        {
+            verb.Text = Loc.GetString("trap-component-set-trap");
+            verb.Act = () =>
+            {
+                if (HandleSetTrap(ent.Owner, localUser))
+                    _doAfter.TryStartDoAfter(doAfterArgs);
+            };
+        }
+
+        args.Verbs.Add(verb);
+    }
+    private void OnTrapDoAfter(Entity<TrapComponent> ent, ref TrapInteractionDoAfterEvent args)
+    {
+        if (args.Cancelled)
+            return;
+
+        var xform = Transform(ent.Owner).Coordinates;
+        _audio.PlayPredicted(ent.Comp.IsArmed ? ent.Comp.SetTrapSound : ent.Comp.DefuseTrapSound, xform, args.User);
+
+        ToggleTrap(ent.Owner, ent.Comp);
+    }
+
+    private void OnStepTriggerAttempt(Entity<TrapComponent> ent, ref StartCollideEvent args)
+    {
+        if(!ent.Comp.IsArmed)
+            return;
+
+        if (_entityWhitelist.IsBlacklistPass(ent.Comp.Blacklist, args.OtherEntity))
+            return;
+
+        if (!TryComp<EnsnaringComponent>(ent.Owner, out var ensnaring))
+            return;
+
+        if (ent.Comp.DurationStun != TimeSpan.Zero && TryComp<StatusEffectsComponent>(args.OtherEntity, out var status))
+        {
+            _stunSystem.TryStun(args.OtherEntity, ent.Comp.DurationStun, true, status);
+            _stunSystem.TryKnockdown(args.OtherEntity, ent.Comp.DurationStun, true, status);
+        }
+
+        var ev = new SharedTriggerEvent(ent.Owner, args.OtherEntity);
+        RaiseLocalEvent(ent.Owner, ev);
+
+        ToggleTrap(ent.Owner, ent.Comp);
+
+        _ensnareableSystem.TryEnsnare(args.OtherEntity, ent.Owner, ensnaring);
+
+        if(_net.IsServer)
+            _audio.PlayPvs(ent.Comp.HitTrapSound, ent.Owner);
+    }
+
+    private void UpdateVisuals(EntityUid uid,TrapComponent? trapComp = null, AppearanceComponent? appearance = null)
+    {
+        if (!Resolve(uid, ref trapComp, ref appearance, false))
+            return;
+
+        _appearance.SetData(uid, TrapVisuals.Visual,
+            trapComp.IsArmed ? TrapVisuals.Armed : TrapVisuals.Unarmed, appearance);
+    }
+
+    private void ToggleTrap(EntityUid uid, TrapComponent comp)
+    {
+        comp.IsArmed = !comp.IsArmed;
+        Dirty(uid,comp);
+        UpdateVisuals(uid);
+
+        if (comp.IsArmed)
+        {
+            _transformSystem.AnchorEntity(uid); //клиентский баг
+        }
+        else
+            _transformSystem.Unanchor(uid);
+
+        var ev = new TrapToggledEvent(comp.IsArmed);
+        RaiseLocalEvent(uid, ev);
+    }
+
+    private bool HandleSetTrap(EntityUid trapEntity, EntityUid user)
```

Comment:
Я так понял это проверка возможности
```suggestion
    private bool CanArmTrap(EntityUid trapEntity, EntityUid user)
```